src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3409,6 +3409,8 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
                     else if (treeInfo.isPartialFunctionMissingParamType(tree)) {
                       if (argPtAlts.exists(ts => isPartialFunctionType(ts._1))) partialFunctionProto(argPtAlts)
                       else functionProto(argPtAlts)
+                    } else if (!treeInfo.isFunctionWithParamType(tree) && argPtAlts.forall { case (t, _) => isFunctionType(t)}) {
```

Comment:
why `argPtAlts.forall { case (t, _) => isFunctionType(t) }`, not also partial / sam function type?

Reply:
Eta-expansion never yields a PartialFunction but we should consider SAM types.